### PR TITLE
Wolf Variants

### PR DESCRIPTION
**Pull request by TheAbsolutionism**
### Description
This PR aims to add wolf variants
---
**Target Minecraft Versions:** 1.21+ 
**Requirements:** none 
**Related Issues:** #6923

### CODE REVIEW

**Review by Moderocky — APPROVED** [review]

File: src/main/java/ch/njol/skript/entity/WolfData.java [hunk 1]
```diff
@@ -18,22 +18,39 @@
...
+			variants = Iterators.toArray(Classes.getExactClassInfo(Wolf.Variant.class).getSupplier().get(), Wolf.Variant.class);
+		}
 	}
 
+
+	private static Object[] variants;
+	@Nullable
+	private Object variant;
 	@Nullable
 	private DyeColor collarColor;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/entity/WolfData.java:55]:
```suggestion
	private @Nullable Object variant;
	private @Nullable DyeColor collarColor;
```

File: src/main/java/ch/njol/skript/entity/WolfData.java [hunk 2]
```diff
@@ -70,11 +91,17 @@ public void set(Wolf entity) {
...
+			variantSet = variant != null ? variant : CollectionUtils.getRandom(variants);
+			entity.setVariant((Wolf.Variant) variantSet);
+		}
 	}
 
 	@Override
 	public boolean match(Wolf entity) {
-		return (angry == 0 || entity.isAngry() == (angry == 1)) && (tamed == 0 || entity.isTamed() == (tamed == 1)) && (collarColor == null ? true : entity.getCollarColor() == collarColor);
+		return (angry == 0 || entity.isAngry() == (angry == 1)) && (tamed == 0 || entity.isTamed() == (tamed == 1)) &&
+			(collarColor == null ? true : entity.getCollarColor() == collarColor) && (variant == null ? true : entity.getVariant() == variant);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/entity/WolfData.java:104]:
```suggestion
			(collarColor == null || entity.getCollarColor() == collarColor) && (variant == null || entity.getVariant() == variant);
```

File: src/test/skript/tests/syntaxes/effects/EffSpawn.sk [hunk 3]
```diff
@@ -10,4 +10,18 @@ test "spawn cats by type" when running minecraft "1.15.2":
...
+	spawn 5 ashen wolves at {_l}
+	assert size of all wolves = 5 with "Size of all wolves is not 5"
+	assert size of all ashen wolves = 5 with "Size of all ashen wolves is not 5"
+	spawn 2 rusty wolves at {_l}
+	assert size of all wolves = 7 with "Size of all wolves is not 7"
+	assert size of all rusty wolves = 2 with "Size of all rusty wolves is not 2"
+	delete all rusty wolves
+	assert size of all wolves = 5 with "Size of all wolves is not 5 after delete 2 rusty wolves"
+	delete all wolves
+	assert size of all wolves = 0 with "Size of all wolves is greater than 0 after all were deleted"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/effects/EffSpawn.sk:27]:
```suggestion
	assert size of all wolves = 0 with "Size of all wolves is greater than 0 after all were deleted"

```

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/entity/WolfData.java [hunk 4]
```diff
@@ -127,7 +157,7 @@ protected boolean deserialize(String s) {
 	public boolean isSupertypeOf(EntityData<?> entityData) {
 		if (entityData instanceof WolfData) {
 			WolfData wolfData = (WolfData) entityData;
-			return (angry == 0 || wolfData.angry == angry) && (tamed == 0 || wolfData.tamed == tamed) && (wolfData.collarColor == collarColor);
+			return (angry == 0 || wolfData.angry == angry) && (tamed == 0 || wolfData.tamed == tamed) && (wolfData.collarColor == collarColor) && (variantsEnabled ? wolfData.variant == variant : true);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/entity/WolfData.java:160]:
```suggestion
			return (angry == 0 || wolfData.angry == angry) && (tamed == 0 || wolfData.tamed == tamed) && (wolfData.collarColor == collarColor) && (!variantsEnabled || wolfData.variant == variant);
```

File: src/test/skript/tests/syntaxes/effects/EffSpawn.sk [hunk 5]
```diff
@@ -10,4 +10,19 @@ test "spawn cats by type" when running minecraft "1.15.2":
...
+
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/effects/EffSpawn.sk:28]:
```suggestion

```

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/effects/EffSpawn.sk (hunk 5 above)
Comment (TheAbsolutionism) [reply src/test/skript/tests/syntaxes/effects/EffSpawn.sk:28]: Didnt you just suggest to add that? or?

File: src/test/skript/tests/syntaxes/effects/EffSpawn.sk (hunk 3 above)
Comment (sovdeeth) [reply src/test/skript/tests/syntaxes/effects/EffSpawn.sk:27]: unresolved

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/classes/data/BukkitClasses.java [hunk 6]
```diff
@@ -1526,6 +1529,24 @@ public String toVariableNameString(EnchantmentOffer eo) {
...
+			 */
+			wolfVariantClassInfo = new ClassInfo<>(WolfData.VariantDummy.class, "wolfvariant");
+		}
+		Classes.registerClass(wolfVariantClassInfo
+			.user("wolf ?variants?")
+			.name("Wolf Variant")
+			.description("Represents the variant of a wolf entity.",
+				"NOTE: Minecraft namespaces are supported, ex: 'minecraft:ashen'.")
+			.since("@VERSION")
+			.requiredPlugins("Minecraft 1.21 or newer")
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/classes/data/BukkitClasses.java:1548]:
```suggestion
			.requiredPlugins("Minecraft 1.21+")
```

**Review by cheeezburga — APPROVED** [review]

**Review by sovdeeth — APPROVED** [review]

**Pull request merged**